streaming/src/main/scala/org/apache/spark/streaming/scheduler/ReceiverTracker.scala

```diff
@@ -247,9 +310,12 @@ class ReceiverTracker(ssc: StreamingContext, skipReceiverLaunch: Boolean = false
 
       if (graceful) {
         val pollTime = 100
+        var elapsedTime = 0
+        def done: Boolean = { receiverInfo.isEmpty && !running }
         logInfo("Waiting for receiver job to terminate gracefully")
-        while (receiverInfo.nonEmpty || running) {
+        while(!done && elapsedTime < TIMEOUT) {
```

Comment:
The `Timeout` logic will change the `stopGracefully` semantics. The `stopGracefully` semantics should be `if ssc.stop(..., stopGracefully = true) returns normally, no data loss will happen`. But after your change, if `ssc.stop(..., stopGracefully = true)` returns normally, the user won't know if everything goes smooth. There is no signal here to help the user understand what happens internally.

I vote for keeping the original codes unchanged.